File: iOS/ai/Data/AnalysisChat.swift
import Foundation

// MARK: - Non-Streaming Chat Methods

extension AnalysisRepository {

    // MARK: - Dispatch

    func sendChatMessage(
        service: AppService,
        apiKey: String,
        model: String,
        messages: [ChatMessage],
        params: ChatParameters
    ) async throws -> String {
        switch service.apiFormat {
        case .anthropic:
            return try await sendChatMessageClaude(service: service, apiKey: apiKey, model: model, messages: messages, params: params)
        case .google:
            return try await sendChatMessageGemini(service: service, apiKey: apiKey, model: model, messages: messages, params: params)
        case .openaiCompatible:
            return try await sendChatMessageOpenAiCompatible(service: service, apiKey: apiKey, model: model, messages: messages, params: params)
        }
    }

    // MARK: - OpenAI Compatible

    private func sendChatMessageOpenAiCompatible(
        service: AppService,
        apiKey: String,
        model: String,
        messages: [ChatMessage],
        params: ChatParameters
    ) async throws -> String {
        // Check Responses API
        if service.usesResponsesApi(model: model) {
            return try await sendChatMessageResponsesApi(service: service, apiKey: apiKey, model: model, messages: messages, params: params)
        }

        var openAiMessages: [OpenAiMessage] = []
        if !params.systemPrompt.isEmpty {
            openAiMessages.append(OpenAiMessage(role: "system", content: params.systemPrompt))
        }
        openAiMessages.append(contentsOf: messages.filter { $0.role != ChatMessage.roleSystem }.map {
            OpenAiMessage(role: $0.role, content: $0.content)
        })

        let request = OpenAiRequest(
            model: model,
            messages: openAiMessages,
            max_tokens: params.maxTokens,
            temperature: params.temperature,
            top_p: params.topP,
            top_k: params.topK,
            frequency_penalty: params.frequencyPenalty,
            presence_penalty: params.presencePenalty,
            search: params.searchEnabled ? true : nil
        )

        let url = ApiClient.chatUrl(for: service)
        let headers = ApiClient.openAiHeaders(apiKey: apiKey)
        let body = try ApiClient.encode(request)

        let (response, _, _): (OpenAiResponse, Int, [String: String]) =
            try await ApiClient.shared.sendRequest(url: url, headers: headers, body: body, responseType: OpenAiResponse.self)

        guard let content = response.choices?.first?.message.content else {
            throw ApiError.streamError("No response content")
        }
        return content
    }

    // MARK: - Responses API

    private func sendChatMessageResponsesApi(
        service: AppService,
        apiKey: String,
        model: String,
        messages: [ChatMessage],
        params: ChatParameters
    ) async throws -> String {
        let systemPrompt = params.systemPrompt.isEmpty ? nil : params.systemPrompt
        let inputMessages = messages.filter { $0.role != ChatMessage.roleSystem }
            .map { OpenAiResponsesInputMessage(role: $0.role, content: $0.content) }

        let request: OpenAiResponsesRequest
        if inputMessages.count == 1, inputMessages.first?.role == "user" {
            request = OpenAiResponsesRequest(model: model, input: .text(inputMessages.first!.content), instructions: systemPrompt)
        } else {
            request = OpenAiResponsesRequest(model: model, input: .messages(inputMessages), instructions: systemPrompt)
        }

        let normalizedBase = ApiClient.normalizeBaseUrl(service.baseUrl)
        let url = "\(normalizedBase)v1/responses"
        let headers = ApiClient.openAiHeaders(apiKey: apiKey)
        let body = try ApiClient.encode(request)

        let (data, statusCode, _) = try await ApiClient.shared.sendRawRequest(url: url, headers: headers, body: body)

        guard (200...299).contains(statusCode) else {
            let errorBody = String(data: data, encoding: .utf8) ?? "Unknown error"
            throw ApiError.httpError(statusCode, errorBody)
        }

        let response = try ApiClient.decode(OpenAiResponsesApiResponse.self, from: data)

        let firstOutput = response.output?.first
        let firstContent = firstOutput?.content
        let content: String? = firstContent?.first(where: { $0.type == "output_text" })?.text
            ?? firstContent?.first(where: { $0.type == "text" })?.text
            ?? firstContent?.compactMap({ $0.text }).first
            ?? response.output?.flatMap({ $0.content ?? [] }).compactMap({ $0.text }).first

        guard let content = content else {
            throw ApiError.streamError("No response content from Responses API")
        }
        return content
    }

    // MARK: - Claude

    private func sendChatMessageClaude(
        service: AppService,
        apiKey: String,
        model: String,
        messages: [ChatMessage],
        params: ChatParameters
    ) async throws -> String {
        let systemPrompt = params.systemPrompt.isEmpty ? nil : params.systemPrompt
        let claudeMessages = messages.filter { $0.role != ChatMessage.roleSystem }
            .map { ClaudeMessage(role: $0.role, content: $0.content) }

        let request = ClaudeRequest(
            model: model,
            max_tokens: params.maxTokens ?? 4096,
            messages: claudeMessages,
            temperature: params.temperature,
            top_p: params.topP,
            top_k: params.topK,
            system: systemPrompt,
            search: params.searchEnabled ? true : nil
        )

        let normalizedBase = ApiClient.normalizeBaseUrl(service.baseUrl)
        let url = "\(normalizedBase)v1/messages"
        let headers = ApiClient.anthropicHeaders(apiKey: apiKey)
        let body = try ApiClient.encode(request)

        let (response, _, _): (ClaudeResponse, Int, [String: String]) =
            try await ApiClient.shared.sendRequest(url: url, headers: headers, body: body, responseType: ClaudeResponse.self)

        guard let content = response.content?.first(where: { $0.type == "text" })?.text else {
            throw ApiError.streamError("No response content from Claude")
        }
        return content
    }

    // MARK: - Gemini

    private func sendChatMessageGemini(
        service: AppService,
        apiKey: String,
        model: String,
        messages: [ChatMessage],
        params: ChatParameters
    ) async throws -> String {
        let contents = messages.filter { $0.role != ChatMessage.roleSystem }
            .map { msg in
                GeminiContent(
                    parts: [GeminiPart(text: msg.content)],
                    role: msg.role == ChatMessage.roleUser ? "user" : "model"
                )
            }

        let systemInstruction = messages.first(where: { $0.role == ChatMessage.roleSystem }).map {
            GeminiContent(parts: [GeminiPart(text: $0.content)])
        } ?? (params.systemPrompt.isEmpty ? nil : GeminiContent(parts: [GeminiPart(text: params.systemPrompt)]))

        let request = GeminiRequest(
            contents: contents,
            generationConfig: GeminiGenerationConfig(
                temperature: params.temperature,
                topP: params.topP,
                topK: params.topK,
                maxOutputTokens: params.maxTokens,
                search: params.searchEnabled ? true : nil
            ),
            systemInstruction: systemInstruction
        )

        let url = ApiClient.geminiGenerateUrl(baseUrl: service.baseUrl, model: model, apiKey: apiKey)
        let headers = ApiClient.geminiHeaders()
        let body = try ApiClient.encode(request)

        let (response, _, _): (GeminiResponse, Int, [String: String]) =
            try await ApiClient.shared.sendRequest(url: url, headers: headers, body: body, responseType: GeminiResponse.self)

        guard let content = response.candidates?.first?.content?.parts.first?.text else {
            throw ApiError.streamError("No response content from Gemini")
        }
        return content
    }
}

File: android/app/src/main/java/com/ai/data/AnalysisChat.kt
package com.ai.data

import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext

/**
 * Send a chat message with conversation history.
 * Returns the assistant's response message.
 */
suspend fun AnalysisRepository.sendChatMessage(
    service: AppService,
    apiKey: String,
    model: String,
    messages: List<com.ai.data.ChatMessage>,
    params: com.ai.data.ChatParameters
): String = withContext(Dispatchers.IO) {
    when (service.apiFormat) {
        ApiFormat.ANTHROPIC -> sendChatMessageClaude(service, apiKey, model, messages, params)
        ApiFormat.GOOGLE -> sendChatMessageGemini(service, apiKey, model, messages, params)
        ApiFormat.OPENAI_COMPATIBLE -> sendChatMessageOpenAiCompatible(service, apiKey, model, messages, params)
    }
}

internal fun AnalysisRepository.convertToOpenAiMessages(messages: List<com.ai.data.ChatMessage>): List<OpenAiMessage> {
    return messages.map { msg -> OpenAiMessage(role = msg.role, content = msg.content) }
}

/**
 * Unified chat method for all OpenAI-compatible providers.
 * Uses OpenAiCompatibleApi with @Url for dynamic chat paths.
 */
internal suspend fun AnalysisRepository.sendChatMessageOpenAiCompatible(
    service: AppService,
    apiKey: String,
    model: String,
    messages: List<com.ai.data.ChatMessage>,
    params: com.ai.data.ChatParameters
): String {
    // Check if this model uses the Responses API (OpenAI gpt-5.x, o3, o4)
    if (usesResponsesApi(service, model)) {
        return sendChatMessageResponsesApi(service, apiKey, model, messages, params)
    }

    val api = ApiFactory.createOpenAiCompatibleApi(service.baseUrl)
    val normalizedBase = if (service.baseUrl.endsWith("/")) service.baseUrl else "${service.baseUrl}/"
    val chatUrl = normalizedBase + service.chatPath

    val openAiMessages = convertToOpenAiMessages(messages)
    val request = OpenAiRequest(
        model = model,
        messages = openAiMessages,
        max_tokens = params.maxTokens,
        temperature = params.temperature,
        top_p = params.topP,
        top_k = params.topK,
        frequency_penalty = params.frequencyPenalty,
        presence_penalty = params.presencePenalty,
        search = if (params.searchEnabled) true else null
    )
    val response = api.createChatCompletion(
        url = chatUrl,
        authorization = "Bearer $apiKey",
        request = request
    )
    if (response.isSuccessful) {
        val content = response.body()?.choices?.firstOrNull()?.message?.content
        return content ?: throw Exception("No response content")
    } else {
        throw Exception("API error: ${response.code()} ${response.message()}")
    }
}

/**
 * Send a non-streaming chat message via OpenAI Responses API (gpt-5.x, o3, o4).
 */
internal suspend fun AnalysisRepository.sendChatMessageResponsesApi(
    service: AppService,
    apiKey: String,
    model: String,
    messages: List<com.ai.data.ChatMessage>,
    params: com.ai.data.ChatParameters
): String {
    val api = ApiFactory.createOpenAiApiWithBaseUrl(service.baseUrl)
    val systemPrompt = messages.find { it.role == "system" }?.content
    val inputMessages = messages
        .filter { it.role != "system" }
        .map { msg -> OpenAiResponsesInputMessage(role = msg.role, content = msg.content) }

    // For single user message, use simple input string; for multi-turn, send full input array.
    // This preserves conversation context for Responses API models.
    val request = if (inputMessages.size == 1 && inputMessages.first().role == "user") {
        OpenAiResponsesRequest(
            model = model,
            input = inputMessages.first().content,
            instructions = systemPrompt
        )
    } else {
        OpenAiResponsesRequest(
            model = model,
            input = inputMessages,
            instructions = systemPrompt
        )
    }

    val response = api.createResponse(
        authorization = "Bearer $apiKey",
        request = request
    )
    if (response.isSuccessful) {
        val body = response.body()
        val content = body?.output?.let { outputs ->
            outputs.firstOrNull()?.content?.firstOrNull { it.type == "output_text" }?.text
                ?: outputs.firstOrNull()?.content?.firstOrNull { it.type == "text" }?.text
                ?: outputs.firstOrNull()?.content?.firstNotNullOfOrNull { it.text }
                ?: outputs.firstOrNull { it.type == "message" }?.content?.firstNotNullOfOrNull { it.text }
                ?: outputs.flatMap { it.content ?: emptyList() }.firstNotNullOfOrNull { it.text }
        }
        return content ?: throw Exception("No response content")
    } else {
        val errorBody = try { response.errorBody()?.string() } catch (_: Exception) { null }
        throw Exception("API error: ${response.code()} ${response.message()} - $errorBody")
    }
}

internal suspend fun AnalysisRepository.sendChatMessageClaude(
    service: AppService,
    apiKey: String,
    model: String,
    messages: List<com.ai.data.ChatMessage>,
    params: com.ai.data.ChatParameters
): String {
    val claudeApi = ApiFactory.createClaudeApiWithBaseUrl(service.baseUrl)
    val claudeMessages = messages
        .filter { it.role != "system" }
        .map { msg -> ClaudeMessage(role = msg.role, content = msg.content) }
    val systemPrompt = messages.find { it.role == "system" }?.content

    val request = ClaudeRequest(
        model = model,
        messages = claudeMessages,
        max_tokens = params.maxTokens ?: 4096,
        temperature = params.temperature,
        top_p = params.topP,
        top_k = params.topK,
        system = systemPrompt,
        search = if (params.searchEnabled) true else null
    )
    val response = claudeApi.createMessage(apiKey = apiKey, request = request)
    if (response.isSuccessful) {
        val content = response.body()?.content?.firstOrNull { it.type == "text" }?.text
        return content ?: throw Exception("No response content")
    } else {
        throw Exception("API error: ${response.code()} ${response.message()}")
    }
}

internal suspend fun AnalysisRepository.sendChatMessageGemini(
    service: AppService,
    apiKey: String,
    model: String,
    messages: List<com.ai.data.ChatMessage>,
    params: com.ai.data.ChatParameters
): String {
    val geminiApi = ApiFactory.createGeminiApiWithBaseUrl(service.baseUrl)
    val contents = messages
        .filter { it.role != "system" }
        .map { msg ->
            GeminiContent(
                parts = listOf(GeminiPart(text = msg.content)),
                role = if (msg.role == "user") "user" else "model"
            )
        }
    val systemInstruction = messages.find { it.role == "system" }?.let {
        GeminiContent(parts = listOf(GeminiPart(text = it.content)))
    }

    val request = GeminiRequest(
        contents = contents,
        generationConfig = GeminiGenerationConfig(
            temperature = params.temperature,
            maxOutputTokens = params.maxTokens,
            topP = params.topP,
            topK = params.topK,
            search = if (params.searchEnabled) true else null
        ),
        systemInstruction = systemInstruction
    )
    val response = geminiApi.generateContent(model, apiKey, request)
    if (response.isSuccessful) {
        val content = response.body()?.candidates?.firstOrNull()?.content?.parts?.firstOrNull()?.text
        return content ?: throw Exception("No response content")
    } else {
        throw Exception("API error: ${response.code()} ${response.message()}")
    }
}
